org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/commands/ProjectCommand.java

```diff
@@ -0,0 +1,271 @@
+			super(String.valueOf(new Date().getTime()), null, false);
+			this.project = project;
+			this.excludeTestCode = excludeTestCode;
+			if (ProjectUtils.isMavenProject(project)) {
+				classpathProvider = MavenRuntimeClasspathProvider.MAVEN_CLASSPATH_PROVIDER;
+				sourcepathProvider = MavenRuntimeClasspathProvider.MAVEN_SOURCEPATH_PROVIDER;
+			} else if (ProjectUtils.isGradleProject(project)) {
+				classpathProvider = GradleClasspathProvider.ID;
+			}
+			// Since MavenRuntimeClasspathProvider will only encluding test entries when:
+			// 1. Launch configuration is JUnit/TestNG type
+			// 2. Mapped resource is in test path.
+			// That's why we use JUnit launch configuration here to make sure the result is right when excludeTestCode is false.
+			// See: {@link org.eclipse.m2e.jdt.internal.launch.MavenRuntimeClasspathProvider#getArtifactScope(ILaunchConfiguration)}
+			this.launchInfo = new JavaLaunchConfigurationInfo(String.format(JAVA_APPLICATION_LAUNCH, excludeTestCode ? 
+					MavenRuntimeClasspathProvider.JDT_JAVA_APPLICATION : MavenRuntimeClasspathProvider.JDT_JUNIT_TEST));
+		}
+
+		@Override
+		public boolean getAttribute(String attributeName, boolean defaultValue) throws CoreException {
+			if (IJavaLaunchConfigurationConstants.ATTR_EXCLUDE_TEST_CODE.equalsIgnoreCase(attributeName)) {
+				return excludeTestCode;
+			}
+
+			return super.getAttribute(attributeName, defaultValue);
+		}
+
+		@Override
+		public String getAttribute(String attributeName, String defaultValue) throws CoreException {
+			if (IJavaLaunchConfigurationConstants.ATTR_PROJECT_NAME.equalsIgnoreCase(attributeName)) {
+				return project.getName();
+			} else if (IJavaLaunchConfigurationConstants.ATTR_CLASSPATH_PROVIDER.equalsIgnoreCase(attributeName)) {
+				return classpathProvider;
+			} else if (IJavaLaunchConfigurationConstants.ATTR_SOURCE_PATH_PROVIDER.equalsIgnoreCase(attributeName)) {
+				return sourcepathProvider;
+			}
+
+			return super.getAttribute(attributeName, defaultValue);
+		}
+
+		@Override
+		protected LaunchConfigurationInfo getInfo() throws CoreException {
+			return this.launchInfo;
+		}
+	}
+
+	private static class JavaLaunchConfigurationInfo extends LaunchConfigurationInfo {
+		public JavaLaunchConfigurationInfo(String launchXml) {
+			super();
+			try {
+				DocumentBuilder parser = DocumentBuilderFactory.newInstance().newDocumentBuilder();
+				parser.setErrorHandler(new DefaultHandler());
+				StringReader reader = new StringReader(launchXml);
+				InputSource source = new InputSource(reader);
+				Element root = parser.parse(source).getDocumentElement();
+				initializeFromXML(root);
+			} catch (ParserConfigurationException | SAXException | IOException | CoreException e) {
+				// do nothing
+			}
+		}
+	}
+
+	public static class ClasspathOptions {
+		public boolean excludingTests;
+	}
+
+	public static class ClasspathResult {
+		public URI projectRoot;
+		public String[] classpaths;
```

Comment:
Classpaths should be scoped: main, test , runtime for maven, Gradle can have custom scopes

Reply:
I took a look at the m2e implementation, seems that m2e will only differentiate `Runtime` & `Test`: https://github.com/eclipse/m2e-core/blob/master/org.eclipse.m2e.jdt/src/org/eclipse/m2e/jdt/internal/DefaultClasspathManagerDelegate.java#L92-L100. (Not quite sure though)

Is there any APIs in m2e which we can leverage to differentiate all the scopes?

Reply:
ping... @fbricon, @snjeza, @testforstephen 

Found it not easy to differentiate all the scopes using `MavenRuntimeClasspathProvider` in `m2e` and `GradleClasspathProvider` in `buildship`.

Can we change the API to:

```
ClasspathQueryOptions = {
    scope: string;
};
```

the scope will only support `runtime` & `test` now. If the input scope is not supported, use `runtime` instead?

Reply:
scope:string is fine I guess

Reply:
Can we have public Map<String, String[]> classpaths?, where the keys will be scopes. I'm not comfortable with scopes being added to seemingly random indexes.